### PR TITLE
Kotlin Conversions

### PR DESCRIPTION
**Pull request by hiqua**
Questions:
1. shall I remove branches such as:
```
            if (SDK_INT >= LOLLIPOP) elevation = dp(1f)
```
since they are always true? Or is there a small chance that we'd go back with the SDK minimum version?

2. I converted org.isoron.uhabits.activities.common.views as well, I didn't see it was crossed, shall I put the java version back? On the other hand now that it's done I guess it's also fine to leave it like this, if my changes leave you enough confidence that I haven't introduced new bugs.

3. Do we delete the package-info.java?

### CODE REVIEW

File: uhabits-android/src/main/java/com/android/datetimepicker/AccessibleTextView.java [hunk 1]
```diff
@@ -27,7 +27,7 @@
 /**
  * Fake Button class, used so TextViews can announce themselves as Buttons, for accessibility.
  */
-public class AccessibleTextView extends TextView {
+public class AccessibleTextView extends androidx.appcompat.widget.AppCompatTextView {
```
Comment (iSoron) [inline uhabits-android/src/main/java/com/android/datetimepicker/AccessibleTextView.java:30]: Please revert all changes to `com.android`, unless there is a strong reason

File: uhabits-android/src/main/java/com/android/datetimepicker/date/TextViewWithCircularIndicator.java [hunk 2]
```diff
@@ -30,7 +30,7 @@
 /**
  * A text view which, when pressed or activated, displays a blue circle around the text.
  */
-public class TextViewWithCircularIndicator extends TextView {
+public class TextViewWithCircularIndicator extends androidx.appcompat.widget.AppCompatTextView {
```
Comment (iSoron) [inline uhabits-android/src/main/java/com/android/datetimepicker/date/TextViewWithCircularIndicator.java:33]: Please revert all changes to `com.android`, unless there is a strong reason

File: uhabits-android/src/main/java/org/isoron/uhabits/activities/AndroidThemeSwitcher.kt [hunk 3]
```diff
@@ -53,9 +53,9 @@ constructor(
         }
     }
 
-    override fun getCurrentTheme(): Theme {
-        return currentTheme
-    }
+    // override fun getCurrentTheme(): Theme {
+    //     return currentTheme
+    // }
```
Comment (iSoron) [inline uhabits-android/src/main/java/org/isoron/uhabits/activities/AndroidThemeSwitcher.kt:58]: If this is no longer useful, please delete

File: uhabits-android/src/main/java/org/isoron/uhabits/activities/common/views/ScrollableChart.kt [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+abstract class ScrollableChart : View, GestureDetector.OnGestureListener, AnimatorUpdateListener {
+    var dataOffset = 0
+        private set
+    private var scrollerBucketSize = 1
+    private var direction = 1
+    private var detector: GestureDetector? = null
+    private var scroller: Scroller? = null
+    private var scrollAnimator: ValueAnimator? = null
+    private var scrollController: ScrollController? = null
```
Comment (iSoron) [inline uhabits-android/src/main/java/org/isoron/uhabits/activities/common/views/ScrollableChart.kt:43]: Please use `lateinit var` instead of nullable type

File: uhabits-android/src/main/java/org/isoron/uhabits/activities/habits/show/views/ScoreCardView.kt [hunk 5]
```diff
@@ -36,7 +36,7 @@ class ScoreCardView(context: Context, attrs: AttributeSet) : LinearLayout(contex
         val androidColor = state.color.toThemedAndroidColor(context)
         binding.title.setTextColor(androidColor)
         binding.spinner.setSelection(state.spinnerPosition)
-        binding.scoreView.setScores(state.scores)
+        binding.scoreView.setScores(state.scores.toMutableList())
```
Comment (iSoron) [inline uhabits-android/src/main/java/org/isoron/uhabits/activities/habits/show/views/ScoreCardView.kt:39]: If possible, please change the type of `ScoreChart.scores` to `List<Score>` instead of changing this line.

File: uhabits-android/src/main/java/org/isoron/uhabits/activities/settings/SettingsFragment.kt [hunk 6]
```diff
@@ -78,7 +78,7 @@ class SettingsFragment : PreferenceFragmentCompat(), OnSharedPreferenceChangeLis
         setResultOnPreferenceClick("bugReport", RESULT_BUG_REPORT)
     }
 
-    override fun onCreatePreferences(bundle: Bundle, s: String) {
+    override fun onCreatePreferences(bundle: Bundle?, s: String?) {
```
Comment (iSoron) [inline uhabits-android/src/main/java/org/isoron/uhabits/activities/settings/SettingsFragment.kt:81]: Why is this needed?

File: uhabits-android/src/main/java/org/isoron/uhabits/preferences/SharedPreferencesStorage.kt [hunk 7]
```diff
@@ -46,32 +46,32 @@ class SharedPreferencesStorage
...
+    override fun putInt(key: String?, value: Int) =
         sharedPrefs.edit().putInt(key, value).apply()
 
-    override fun putLong(key: String, value: Long) =
+    override fun putLong(key: String?, value: Long) =
         sharedPrefs.edit().putLong(key, value).apply()
 
-    override fun putString(key: String, value: String) =
+    override fun putString(key: String?, value: String?) =
         sharedPrefs.edit().putString(key, value).apply()
```
Comment (iSoron) [inline uhabits-android/src/main/java/org/isoron/uhabits/preferences/SharedPreferencesStorage.kt:75]: Why are these changes needed? Please revert if not required.

File: uhabits-android/src/main/java/org/isoron/uhabits/tasks/AndroidTaskRunner.kt [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        override fun onPostExecute(aVoid: Void?) {
+            if (isCancelled) return
+            task.onPostExecute()
+            activeTasks.remove(this)
+            taskToAsyncTask.remove(task)
+            for (l in listeners) l.onTaskFinished(task)
+        }
+
+        override fun onPreExecute() {
+            // isCancelled = task.isCanceled
```
Comment (iSoron) [inline uhabits-android/src/main/java/org/isoron/uhabits/tasks/AndroidTaskRunner.kt:78]: If this is not needed, please delete

File: uhabits-android/src/main/java/org/isoron/uhabits/widgets/ScoreWidget.kt [hunk 9]
```diff
@@ -52,7 +52,7 @@ class ScoreWidget(
             setIsTransparencyEnabled(true)
             setBucketSize(viewModel.bucketSize)
             setColor(habit.color.toThemedAndroidColor(context))
-            setScores(viewModel.scores)
+            setScores(viewModel.scores.toMutableList())
```
Comment (iSoron) [inline uhabits-android/src/main/java/org/isoron/uhabits/widgets/ScoreWidget.kt:55]: See previous comment about this

File: uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/io/HabitsCSVExporter.kt [hunk 10]
```diff
@@ -77,7 +77,7 @@ class HabitsCSVExporter(
     }
 
     private fun sanitizeFilename(name: String): String {
-        val s = name.replace("[^ a-zA-Z0-9\\._-]+".toRegex(), "")
+        val s = name.replace("[^ a-zA-Z0-9._-]+".toRegex(), "")
```
Comment (iSoron) [inline uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/io/HabitsCSVExporter.kt:80]: Why is this needed?

File: uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/models/HabitList.kt [hunk 11]
```diff
@@ -139,31 +108,28 @@ public ModelObservable getObservable()
...
-    }
+    abstract fun indexOf(h: Habit): Int
+    val isEmpty: Boolean
+        get() = size() == 0
 
     /**
      * Removes the given habit from the list.
-     * <p>
+     *
+     *
```
Comment (iSoron) [inline uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/models/HabitList.kt:118]: Please replace by a single empty line. The same is true for the other comments in this class.

File: uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/models/HabitList.kt [hunk 12]
```diff
@@ -217,53 +181,37 @@ public void update(@NonNull Habit habit)
...
-        BY_COLOR_ASC,
-        BY_COLOR_DESC,
-        BY_SCORE_ASC,
-        BY_SCORE_DESC,
-        BY_STATUS_ASC,
-        BY_STATUS_DESC,
-        BY_POSITION
+    abstract fun resort()
+    enum class Order {
+        BY_NAME_ASC, BY_NAME_DESC, BY_COLOR_ASC, BY_COLOR_DESC, BY_SCORE_ASC, BY_SCORE_DESC, BY_STATUS_ASC, BY_STATUS_DESC, BY_POSITION
```
Comment (iSoron) [inline uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/models/HabitList.kt:215]: Please place these in separate lines.

File: uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/models/Timestamp.kt [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+        if (this === other) return true
+        if (other == null || javaClass != other.javaClass) return false
+        val timestamp = other as Timestamp
+        return EqualsBuilder()
+            .append(unixTime, timestamp.unixTime)
+            .isEquals
+    }
+
+    override fun hashCode(): Int {
+        return HashCodeBuilder(17, 37).append(unixTime).toHashCode()
```
Comment (iSoron) [inline uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/models/Timestamp.kt:62]: I think these are no longer required if we make it into a `data class`

File: uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/models/memory/MemoryHabitList.kt [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+            Comparator { h1: Habit, h2: Habit -> statusComparatorDesc.compare(h2, h1) }
+        if (order === Order.BY_POSITION) return positionComparator
+        if (order === Order.BY_NAME_ASC) return nameComparatorAsc
+        if (order === Order.BY_NAME_DESC) return nameComparatorDesc
+        if (order === Order.BY_COLOR_ASC) return colorComparatorAsc
+        if (order === Order.BY_COLOR_DESC) return colorComparatorDesc
+        if (order === Order.BY_SCORE_DESC) return scoreComparatorDesc
+        if (order === Order.BY_SCORE_ASC) return scoreComparatorAsc
+        if (order === Order.BY_STATUS_DESC) return statusComparatorDesc
+        if (order === Order.BY_STATUS_ASC) return statusComparatorAsc
```
Comment (iSoron) [inline uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/models/memory/MemoryHabitList.kt:162]: Please replace by `when`

File: uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/ui/screens/habits/list/ListHabitsBehavior.kt [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+            return
+        }
+        screen.showConfirmInstallSyncKey {
+            prefs.enableSync(syncKey, encryptionKey)
+            screen.showMessage(Message.SYNC_ENABLED)
+        }
+    }
+
+    enum class Message {
+        COULD_NOT_EXPORT, IMPORT_SUCCESSFUL, IMPORT_FAILED, DATABASE_REPAIRED, COULD_NOT_GENERATE_BUG_REPORT, FILE_NOT_RECOGNIZED, SYNC_ENABLED, SYNC_KEY_ALREADY_INSTALLED
```
Comment (iSoron) [inline uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/ui/screens/habits/list/ListHabitsBehavior.kt:126]: Please put these in separate lines

File: uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/ui/screens/habits/list/ListHabitsSelectionMenuBehavior.kt [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+import javax.inject.Inject
+
+class ListHabitsSelectionMenuBehavior @Inject constructor(
+    private val habitList: HabitList,
+    private val screen: Screen,
+    private val adapter: Adapter,
+    var commandRunner: CommandRunner
+) {
+    fun canArchive(): Boolean {
+        for ((_, _, _, _, isArchived) in adapter.selected) if (isArchived) return false
```
Comment (iSoron) [inline uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/ui/screens/habits/list/ListHabitsSelectionMenuBehavior.kt:40]: Please replace `(_,_,_,)` by something more readable. Same for the other occurrences in this file.

File: uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/ui/screens/habits/list/ListHabitsSelectionMenuBehavior.kt [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    fun onArchiveHabits() {
+        commandRunner.run(ArchiveHabitsCommand(habitList, adapter.selected))
+        adapter.clearSelection()
+    }
+
+    fun onChangeColor() {
+        val selected = adapter.selected
+        val (color) = selected[0]
```
Comment (iSoron) [inline uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/ui/screens/habits/list/ListHabitsSelectionMenuBehavior.kt:60]: Please inline `selected`

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/database/RepositoryTest.kt [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+        db.execute(
+            "insert into tests(id, color_number, name, score) " +
+                "values (10, 20, 'hello', 8.0)"
+        )
+        val record = repository.find(10L)
+        Assert.assertNotNull(record)
+        MatcherAssert.assertThat(record!!.id, IsEqual.equalTo(10L))
+        MatcherAssert.assertThat(record.color, IsEqual.equalTo(20))
+        MatcherAssert.assertThat(record.name, IsEqual.equalTo("hello"))
+        MatcherAssert.assertThat(record.score, IsEqual.equalTo(8.0))
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/database/RepositoryTest.kt:61]: Please use static imports for matchers and assertions

**Review by iSoron — CHANGES_REQUESTED** [review]
Thank you for the pull request, @hiqua! I made only some minor suggestions. Since the other two pull requests seem to depend on this one, I will convert them into "draft" until this one gets merged.

**Comment by iSoron** [comment]
Answering your questions:

1. Please feel free to remove the branches. I am not planning to revert the changes to min SDK version. We also don't need a separate directory for `v21` resources anymore.
2. I crossed it out to avoid work, but since you've converted them already, I can merge it.
3. Feel free to delete them.

File: uhabits-android/src/main/java/org/isoron/uhabits/activities/settings/SettingsFragment.kt (hunk 6 above)
Comment (hiqua) [reply uhabits-android/src/main/java/org/isoron/uhabits/activities/settings/SettingsFragment.kt:81]: Preferences actually don't work with the last commit on `dev` without these changes. I'll extract this fix from the commit to make it more explicit. The arguments from `onCreatePreferences` in PreferenceFragmentCompat can be null.

File: uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/io/HabitsCSVExporter.kt (hunk 10 above)
Comment (hiqua) [reply uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/io/HabitsCSVExporter.kt:80]: Not needed, it's just a redundant escape. Shall I revert this change?

**Comment by hiqua** [comment]
> Since the other two pull requests seem to depend on this one, I will convert them into "draft" until this one gets merged.

Makes sense! I'll rebase as soon as this one is merged in.

**Review by iSoron — APPROVED** [review]

**Comment by iSoron** [comment]
Merged. Thanks again, @hiqua!
